docs: Livestream Tutorial (#509)
GetStream/stream-video-js: packages/react-sdk/docusaurus/docs/React/02-tutorials/03-livestream.mdx

---
title: Livestream Tutorial
description: How to build a live stream experience using Stream's React Video SDKs
---

## Quickstart

> In this quickstart, we'll build a live streaming experience that's similar to Twitch.

Reviewer: ```suggestion
In this quickstart, we'll build a live-streaming experience that is similar to Twitch.
```

- The live stream will run on Stream's Edge network of servers around the world

> - You can use ultra-low latency (WebRTC) based live streaming or HLS (slight delay, but better buffering)

Reviewer: ```suggestion
- We can use ultra-low latency ([WebRTC](https://webrtc.org/)) based live streaming or [HLS](https://en.wikipedia.org/wiki/HTTP_Live_Streaming) (slight delay, but better buffering)
```

> - We'll show how to publish to the live stream from your browser

Reviewer: ```suggestion
- We will go through the process of publishing the live stream from the browser
```

> - Stream enables you to scale to millions of viewers

Reviewer: We could explain a bit more that this is done automagically and the user doesn't do anything. Might be misleading though, because costs will obv. increase....not sure if it makes sense.

```suggestion
- Stream enables us to automatically scale to millions of viewers (no actions required from your side)
```

Author: I think this is an implicit fact. I'll take your suggestion here :)

> - You can have multiple active participants in the live stream

Reviewer: I think the switch between "we can" and "you will" is a bit confusing to read. So I switched this to use we in all cases.

```suggestion
- We even show how to have multiple active participants in the live stream
```

> The app that we are going to build would work in two modes: Host (presenter) and Viewer (audience).

Reviewer: ```suggestion
The app that we are going to build will work in two modes: Host (presenter) and Viewer (audience).
```

Reviewer 2: It occurred to me we could provide a description about how the app will work, e.g.:

In this app the host will initiate the livestream meanwhile the viewers are waiting for the livestream to start...

The host will be able to publish the live stream and the viewers will be able to watch the live stream.

> If you want to see a fully-fledged example, check out this [sample app.](https://github.com/GetStream/stream-video-js/tree/main/sample-apps/react/livestream-app)

Reviewer: ```suggestion
If you want to see a full-blown example, check out this [sample app.](https://github.com/GetStream/stream-video-js/tree/main/sample-apps/react/livestream-app)
```

Reviewer 2: I am thinking, that with so many apps we provide in our tutorials, we could introduce an element / banner that would indicate clearly that this is the accompanying sample app / source code. WDYT? Speaking about being more visual and make it easier for readers to find it / determine, that accompanying src code exists.

Author: @MartinCupela, I think this is a good idea. I'm sure other SDKs can benefit from such elements as they also provide sample apps like ours.
I'm not sure how much effort would be required to develop such a component, therefore, I'd leave it out for better days :)

Let's get started.

## Step 1 - Preparations

Before going into the code, let's make sure we have everything we need.

- Add Stream's React Video SDK to your project:
  - `yarn install @stream-io/video-react-sdk` or
  - `npm install @stream-io/video-react-sdk`
- Setup a Stream account at [GetStream's Dashboard](https://dashboard.getstream.io) or use your existing one

>   - We will need your API key and secret

Reviewer: That sounds a bit like: GIVE US YOUR DATA!!!1!!11!

```suggestion
  - We will need an API key and secret to run the project
```

## Step 2 - Setup the host mode

> Hosts should be communicating with each other in real-time. For that purpose, we would use the standard Stream Video Call component, based on WebRTC.

Reviewer: Do you think it makes sense to use *should* and *would* here? 

It sounds a bit theoretical whereas we're indeed creating a real application where everything *WILL* work

Reviewer 2: I would change `Stream Video Call component` to a real component name.

Author: Agreed, I'll update accordingly.

Author: Updated.

:::note
In this tutorial, we will skip the details on how a `StreamVideoClient` can be set up. You can read more about it on the following page:
[Authenticating your Users](../../guides/01-authentication/)

> :::

Reviewer: Bringing some color into play and even linking to other content pages for viewers to go to...Oliver, you are a real Pro! 🚀👑

> In the snippet below, we create a connection to the Stream Edge server for the live stream host, we join the call and we publish our audio and video to the live stream.

Reviewer: I like this explanation a lot. Clear, precise, and to the point. Awesome!

> In this example, we use the predefined `<PaginatedGridLayout>` for the live stream.

Reviewer: Do we have more info about the existing layouts somewhere? Would be a chance to link those.

Author: We should dedicate a chapter for them in `Video UI components`.

> In case you want to have a more custom layout, you can build your own. See [Building Custom Layouts.](../ui-cookbook/participant-list/)

Reviewer: ```suggestion
In case you want to have a more custom layout, you can build your own. See our article about [Building Custom Layouts.](../ui-cookbook/participant-list/)
```

Reviewer 2: ```suggestion
If you prefer a more customized layout, you can build your own. See our article about [Building Custom Layouts.](../ui-cookbook/participant-list/)
```

> ```tsx

Reviewer: I am in no way able to judge the code itself. I feel like it's very long though and might be hard to follow in one bit. Do we have any chances of breaking things up a bit to not make it seem so long?

If not and it becomes unclear or harder to follow, then just keep it as is!

Very naive idea:
We could break it up and extract the `BackstageUI` component into a separate code block, quickly explaining what it does and what it is for...

Author: @DaemonLoki I couldn't make the snippet shorter than this without losing important information. 

@MartinCupela I intentionally excluded a screenshot, as this tutorial doesn't build a UI, instead, it shows how a UI can be built.
At the end of the tutorial, the reader can see a live preview of our sample app, and play with it.

Reviewer 2: I propose to add the links to the preview also at the beginning. I think it can help the understanding what is then read, because the reader can associate what was seen and experienced with the actual code. At least for me it always worked better.

Author: The link to the source code is also available on the top of the tutorial but it isn't made that highlighted as the links at the end.

Reviewer 2: What if the Tip banner from the bottom was moved to the top?

Author: I'm not sure if we want to promote the App first. If we drag the readers to the app first, they might not know what they are looking at.
The guide shows an approach to how a lives stream app can be built. Then we provide a sample implementation.

@DaemonLoki, what is your take on this? What does more sense?

  useCall,
  useCallCallingState,
  useCreateStreamVideoClient,
  useIsCallBroadcastingInProgress,
  useMediaDevices,
  CancelCallButton,
  PaginatedGridLayout,
  StreamMeeting,
  StreamVideo,
  ToggleAudioPublishingButton,
  ToggleCameraPublishingButton,
} from '@stream-io/video-react-sdk';

// ... API key, provider, and user setup

export const HostApp = () => {
  const client = useCreateStreamVideoClient({
    apiKey,
    tokenOrProvider,
    user,
  });

  return (
    <StreamVideo client={client}>
      <StreamMeeting
        callType="livestream"
        callId="some-random-call-id"
        autoJoin={true}
        data={{
          create: true,
          // optionally, invite other hosts to the live stream
          data: {
            members: [...],
          }
        }}
      >
        <BackstageUI />
      </StreamMeeting>
    </StreamVideo>
  )
}

const BackstageUI = () => {
  const call = useCall();
  const callState = useCallCallingState();
  // useMediaDevices() hook provides you quick access to the Browser's MediaDevices API
  // and a few helper methods to manage your audio and video streams.
  const { publishVideoStream, publishAudioStream } = useMediaDevices();

>   useEffect(() => {

Reviewer 2: ```suggestion

  useEffect(() => {
```

    if (!call) return;
    if (callState === CallingState.JOINED) {
      // once the call is joined, publish the host's audio and video
      publishVideoStream();
      publishAudioStream();
    }
  }, [call, callState, publishAudioStream, publishVideoStream]);

>   return (

Reviewer 2: ```suggestion

  return (
```

    </>
  );
};

const BackstageControls = () => {
  return (
    <div className="backstage-controls">
      <ToggleAudioPublishingButton />
      <ToggleCameraPublishingButton />
      <CancelCallButton />
      // The implementation of this button is shown in the next step
      <ToggleLivestreamButton />
    </div>
  );
};

```

With this done, we have a fully functional livestream experience for the host.
The hosts can join the live stream, publish their audio and video, and invite other hosts to the live stream.

:::note
When the call is started, the live streaming isn't enabled for the viewers yet. 
We will see how to enable it in the next step.
::: 

## Step 3 - Enable the live stream for the viewers

For simplicity, we are going to implement an additional button in the `BackstageControls` component that will enable the live stream for the viewers.

The button should be able to toggle the live stream on and off. For that purpose, we are going to use two utility hooks from the SDK: 
- [`useCall`](../04-call-engine/hooks-and-contexts.md#usecall) and
- [`useIsCallBroadcastingInProgress`](../04-call-engine/hooks-and-contexts.md#useiscallbroadcastinginprogress)

> The `useCall` hook provides you with the current call object. This is useful when you want to access the call object from a component that is deep in the component tree with the `<StreamMeeting />` component as a predecessor.

Reviewer: Let's keep it as "we" all the way through, switching between it and "you" is a bit confusing I think.

```suggestion
The `useCall` hook provides us with the current call object. This is useful when we want to access the call object from a component that is deep in the component tree with the `<StreamMeeting />` component as a predecessor.
```

> On the other hand, the `useIsCallBroadcastingInProgress` hook provides you with the current broadcasting state of the call. This is useful when you want to show the current state of the live stream or build custom UI components or flows based on this call state.

Reviewer: ```suggestion
On the other hand, the `useIsCallBroadcastingInProgress` hook gives us the current broadcasting state of the call. This is helpful when we want to show the current state of the live stream or build custom UI components or flows based on this call state.
```

The [`call`](../04-call-engine/Call.md) instance itself provides two methods for starting and stopping the live stream: 
- [`call.startBroadcasting()`](../04-call-engine/Call.md#startbroadcasting) and 
- [`call.stopBroadcasting()`](../04-call-engine/Call.md#stopbroadcasting)

Knowing all of this, let's write the code for the button:

```tsx
const ToggleLivestreamButton = () => {
  const call = useCall();
  const isBroadcasting = useIsCallBroadcastingInProgress();
  return (
    <Button onClick={() => {
        if (isBroadcasting) {
          call.stopBroadcasting();
        } else {
          call.startBroadcasting();
        }
    }}>
      {isBroadcasting ? 'Stop Stream' : 'Start Stream'}
    </Button>
  )
};

```

:::info
Starting and stopping a live stream for the viewers is an operation that takes a few seconds to complete.
It is a good practice to show a loading indicator while the operation is in progress.

> There are two ways how you can do that:

Reviewer: ```suggestion
There are two ways how we can do that:
```

> - have a `useEffect` that would run whenever the we have initiated a change in the broadcasting state of the call until the value of `isBroadcasting` toggles.

Reviewer 2: ```suggestion
- have a `useEffect` that would run whenever we have initiated a change in the broadcasting state of the call until the value of `isBroadcasting` toggles.
```

:::

## Step 4 - Setup the viewer mode

The viewer mode is very similar to the host mode. The only difference is that the viewer doesn't have to publish their audio and video to the live stream.

A viewer can join a live stream in two ways:
- WebRTC-based live streaming (ultra-low latency), by using Stream Edge servers
- HLS-based live streaming (slight delay, but better buffering), by using Stream CDN

### WebRTC-based live streaming

For the WebRTC, we are going to use the same `<StreamMeeting />` component as in the host mode. The only difference is that we are going to use an anonymous user for the viewer.

:::caution
Anonymous users can't establish a WebSocket connection to our backend, therefore they are unable to receive any events from the call.
This means that the viewer won't know when the call starts or stops broadcasting.

To circumvent this limitation, the Viewer app should be able to be notified by the Host app or some other backend service when the call starts or stops broadcasting.

> Alternatively, you can keep polling the backend for the current state of the call. See [Querying Calls.](../02-guides/06-querying-calls.mdx)

Reviewer: ```suggestion
Alternatively, you can keep polling the backend for the current state of the call. For more details on this, see [Querying Calls.](../02-guides/06-querying-calls.mdx)
```

  const [activeCall] = client.queryCalls({...});
  return (
    <StreamVideo client={client}>
      <StreamMeeting 
        call={activeCall} 
        // we don't want to join this call, as that would switch to WebRTC-based live streaming.
        // we are going to use HLS.js to play the live stream
        autoJoin={false}
      >
        <HLSLivestreamUI />
      </StreamMeeting>
    </StreamVideo>
  )
}

export const HLSLivestreamUI = () => {
  // The HLS playlist URL is in call's metadata
  // we'll use the `useCallMetadata()` hook to get it
  const { hls_playlist_url } = useCallMetadata();

  const [isPlaying, setIsPlaying] = useState(false);
  const videoRef = useRef<HTMLVideoElement>(null);
  useEffect(() => {
    if (!call || !hls_playlist_url) return;
  
    const hls = new HLS();
    hls.loadSource(hls_playlist_url);
    hls.attachMedia(videoRef.current);
  }, [call]);

  return (
    <>
      <ViewerHeader />
      // this is the video that HLS.js would use to playback the stream.
      <video autoPlay playsInline ref={videoRef} onPlay={() => setIsPlaying(true)} />
      
      {!isPlaying && <LoadingIndicator />}
      <ViewerControls />
    </>
  );
};
```

> Having this done, we should have a fully functional live streaming experience for the viewers based on HLS.

Reviewer: Let's be bold here, we SHOULDN'T have it but in fact WE REALLY DO!!!💪🏻

```suggestion
Having done this, we now have a fully functional live-streaming experience for the viewers based on HLS.
```

Author: Indeed! Fair point :)

:::tip
For simplicity, a few things are omitted from the code snippet above. For the full-blown and working example,
check this Livestreaming sample app:
- [Livestream App source code](https://github.com/GetStream/stream-video-js/tree/main/sample-apps/react/livestream-app)
- [Livestream App live preview](https://video-react-livestream-app.vercel.app/)

> :::

Reviewer: Even suggesting other content to users. Hat tip, Oliver!

This means that the viewer won't know when the call starts or stops broadcasting.

To circumvent this limitation, the Viewer app should be able to be notified by the Host app or some other backend service when the call starts or stops broadcasting.
Alternatively, you can keep polling the backend for the current state of the call. For more details on this, see [Querying Calls.](../02-guides/06-querying-calls.mdx)
:::

```tsx
import {
  useCall,
  useCallCallingState,
  LoadingIndicator,
  PaginatedGridLayout,
  StreamMeeting,
  StreamVideo,
} from '@stream-io/video-react-sdk';

// ... API key, provider, and user setup

export const ViewerApp = () => {
  const client = useCreateStreamVideoClient({
    apiKey,
    tokenOrProvider,
    user,
    isAnonymous: true,
  });

  const [activeCall, setActiveCall] = useState<Call>();
  useEffect(() => {
    if (!callId) return;
    client
      .queryCalls({
        filter_conditions: {
          cid: { $in: [`livestream:${callId}`] },
        },
        sort: [{ field: 'cid', direction: 1 }],
      })
      .then(({ calls }) => {
        const [call] = calls;
        if (call) {
          setActiveCall(call);
        }
      });

>   }, [client, callId]);

Reviewer 2: It is not entirely clear to me, from where the `callId` comes. That could happen also to other readers.

Author: I've added more information.

  return (
    <StreamVideo client={client}>
      <StreamMeeting call={activeCall} autoJoin={true}>
        <WebRTCLivestreamUI />
      </StreamMeeting>
    </StreamVideo>
  )
}

const WebRTCLivestreamUI = () => {
  const call = useCall();
  const isJoined = useCallCallingState() === CallingState.JOINED;
  return (
    <>
      <ViewerHeader />
      // use any layout from the SDK or build your own
      <PaginatedGridLayout />

      // show a loading indicator while the call is joining
      {!isJoined && <LoadingIndicator />}
      <ViewerControls />
    </>
  );
};

```

> ### HLS-based live streaming

Reviewer 2: As the HLS example uses new hook `useCallMetadata` I would mention it and add a link to the generated docs.

Author: Added.

For the HLS scenario, we are going to use one very popular library called `hls.js`.
HLS.js is a JavaScript library that implements an HTTP Live Streaming client. It relies on HTML5 video and MediaSource Extensions for playback.
You can read more about HLS.js [here.](https://github.com/video-dev/hls.js/)

For the HLS-based live streaming, we are going to use a few components from the Stream Video SDK for React but, instead using a pre-built layout,
we are going to set up the HLS.js library and use it to play the live stream.

Compared to the WebRTC example, we would just replace the "UI" component with the following:

```tsx
import { ... } from '@stream-io/video-react-sdk';
import HLS from 'hls.js';

export const ViewerApp = () => {
  // pseudo-code for simplicity, see the previous example for the full code
  const client = useCreateStreamVideoClient({...});
  const [activeCall] = client.queryCalls({...});
  return (
    <StreamVideo client={client}>
      <StreamMeeting 
        call={activeCall} 
        // we don't want to join this call, as that would switch to WebRTC-based live streaming.
        // we are going to use HLS.js to play the live stream

>         autoJoin={false}

Reviewer 2: ```suggestion
// highlight-next-line
        autoJoin={false}
```

      >

>         <HLSLivestreamUI />

Reviewer 2: ```suggestion
// highlight-next-line
        <HLSLivestreamUI />
```
